### PR TITLE
Code Refactor and Reprocessing Functionality

### PR DESCRIPTION
**Pull request by LukeSeifert**
This PR includes a refactor of the code from the `flow-tests` directory, proper incorporation of Serpent2 built-in flow functionality, and several different approximations.

The `Notes` directory contains a Jupyter notebook with information on how the different reprocessing constant approximations are generated.

The `msbr_compare` directory is where the refactor and new functionality is located. The `readme` contains most of the information on the overall purpose of the scripts, as well as what each individual script is meant to do.

### CODE REVIEW

File: msr-flows/Notes/type_1_flow.ipynb
Comment (yardasol) [inline msr-flows/Notes/type_1_flow.ipynb:9]: Use expanded acronym on first use.
```suggestion
    "#### Only suitable for steady state (SS) calculations \n",
```

File: msr-flows/Notes/type_1_flow.ipynb
Comment (yardasol) [inline msr-flows/Notes/type_1_flow.ipynb:11]: Here's some formatting and flow suggestions to enhance clarity. I'm not quite sure if my edits here reflect what you're trying to convey. If they don't, you can ignore this suggestion.
```suggestion
    "The _linear growth approximation_ is an approach which assumes that over some time step X, the growth is approximately linear. This means that the removal constant must be calculated such that the number of atoms at time X is the same as in the SaltProc calculation at time X. As long as accuracy can be maintained to some position past where the SaltProc value is determined, then this method can be used to reduce the computational time of SaltProc by making a series of linear approximations.\n",
```

File: msr-flows/Notes/type_1_flow.ipynb
Comment (yardasol) [inline msr-flows/Notes/type_1_flow.ipynb:13]: This is a great figure for illustrating what the LGA does! Good job.

File: msr-flows/Notes/type_1_flow.ipynb
Comment (yardasol) [inline msr-flows/Notes/type_1_flow.ipynb:15]: Good job explaining the steps of this algorithm!

File: msr-flows/Notes/type_1_flow.ipynb
Comment (yardasol) [reply msr-flows/Notes/type_1_flow.ipynb:9]: tagging @nsryan2 for another opinion

File: msr-flows/Notes/type_1_flow.ipynb
Comment (yardasol) [reply msr-flows/Notes/type_1_flow.ipynb:11]: tagging @nsryan2 for another opinion

File: msr-flows/Notes/type_1_flow.ipynb
Comment (yardasol) [inline msr-flows/Notes/type_1_flow.ipynb:25]: Flow + formatting edit for enhanced clarity. Same quantifier on your intended message applies to this suggestion.
```suggestion
    "2. Plug the linear growth term into the Bateman equation and then include reprocessing with the condition that it must match the SaltProc value at a given time. This is done by calculating waste atoms and solving with `scipy.optimize` root function.     \n",
```

Comment (yardasol) [reply msr-flows/Notes/type_1_flow.ipynb:25]: tagging @nsryan2 for another opinion

File: msr-flows/Notes/type_1_flow.ipynb
Comment (yardasol) [inline msr-flows/Notes/type_1_flow.ipynb:48]:
```suggestion
    "#### Suitable for all calculations\n",
```

File: msr-flows/Notes/type_1_flow.ipynb
Comment (yardasol) [inline msr-flows/Notes/type_1_flow.ipynb:52]: >the half-lives are determined from the cycle times from Robertson et al. report on the MSBR that Andrei cites in the MSBR example.

This is a bit vague. Consider using the path to the file where the citation to the Robertson paper exists. Or if possible, include a link to the paper directly.

File: msr-flows/Notes/type_1_flow.ipynb
Comment (yardasol) [inline msr-flows/Notes/type_1_flow.ipynb:106]: It's not clear to me what this link is for. Perhaps you could user hypertext to clarify your intent?

File: msr-flows/msbr_compare/readme.md
Comment (yardasol) [inline msr-flows/msbr_compare/readme.md:3]: excellent description

File: msr-flows/msbr_compare/readme.md
Comment (yardasol) [inline msr-flows/msbr_compare/readme.md:11]:
```suggestion
### `run_user_input.py`
```

File: msr-flows/msbr_compare/readme.md
Comment (yardasol) [inline msr-flows/msbr_compare/readme.md:8]:
```suggestion
### `user_input.py`
```

File: msr-flows/msbr_compare/readme.md
Comment (yardasol) [inline msr-flows/msbr_compare/readme.md:14]:
```suggestion
### `serpent_input.py`
```

File: msr-flows/msbr_compare/readme.md
Comment (yardasol) [inline msr-flows/msbr_compare/readme.md:17]:
```suggestion
### `serpent_calculations.py`
```

File: msr-flows/msbr_compare/readme.md
Comment (yardasol) [inline msr-flows/msbr_compare/readme.md:20]:
```suggestion
### `serpent_output.py`
```

File: msr-flows/msbr_compare/readme.md
Comment (yardasol) [inline msr-flows/msbr_compare/readme.md:23]:
```suggestion
### `misc_funcs.py`
```

File: msr-flows/msbr_compare/readme.md
Comment (yardasol) [inline msr-flows/msbr_compare/readme.md:26]:
```suggestion
### `saltproc_read.py`
```

File: msr-flows/msbr_compare/readme.md
Comment (yardasol) [inline msr-flows/msbr_compare/readme.md:39]: Consider making these issues in this repo

File: msr-flows/msbr_compare/saltproc_read.py
Comment (yardasol) [inline msr-flows/msbr_compare/saltproc_read.py:24]: Consider adding description for `acelib_path`

File: msr-flows/msbr_compare/saltproc_read.py
Comment (yardasol) [inline msr-flows/msbr_compare/saltproc_read.py:6]: consider adding `Paremeters` section to describe `isotope` and `lib_isos`

File: msr-flows/msbr_compare/serpent_output.py
Comment (yardasol) [inline msr-flows/msbr_compare/serpent_output.py:1]: I think `pyne` also has a serpent module. But you might have already tried using that.

File: msr-flows/readme.md
Comment (yardasol) [inline msr-flows/readme.md:5]: I assume these deletions are to reflect the refactor?

**Review by yardasol — CHANGES_REQUESTED** [review]
Hi @LukeSeifert,
Thanks for your PR! This repo hasn't been updated in a while so I'm glad to see you're giving it the TLC it deserves.

## General feedback

Overall, I really like your level of detail and documentation. There are some changes that I suggest you consider making before I approve this PR.

### PR needs linked issues
In the [page on making PR's on our website](https://arfc.github.io/manual/guides/pull_requests):
> PRs made in the ARFC organization should:
>   
  .
  .
  .
>    - Link to an issue

During the process of writing this review, I've looked at the description of this PR several times and had questions about what the described solutions are fixing and why this PR contains the things it does (e.g. What about the previous incorporation of Serpent2 build-in flow functionality was improper? Why does the code in the `flow-tests` directory need refactoring?). Knowing what issues this PR is resolving would help me understand the PR description. I've skimmed the existing issues in this repo and it looks like you probably have ones you can link here that would answer those questions.

### Merging from `master`
This may not be relevant but I'll say it anyways just in case it turns out to be. I see you are merging from your fork's `master` branch. While this repository doesn't seem to have any contributing guidelines, but if we take a look at [Moltres' `CONTRIBUTING.md` file](https://github.com/arfc/moltres/blob/devel/CONTRIBUTING.md):
>Keep your own "master" and "devel" branches in sync with the main repository's "master" and "devel" branches. Specifically, do not push your own commits directly to your "master" and "devel" branches.


My interpretation of this is that you can always create branches off of your forks `master` with the most recent working version of the code to develop multiple features independently (hopefully) without conflicting. I'm not clear on whether or not this general workflow applies to _all_ repos in this group, but regardless I thought I'd point this out just as something for you to consider.

## Review comments:
In addition to my feedback in line:
- Looks like you have some PEP8 issues. You can easily fix these by executing the following in the directory where you have your `.py` files:
   ```
   autopep8 --in-place *.py
   git add *.py
   git commit -m "pep8 fixes"
   git push
   ```
   rinse and repeat till @pep8speaks stops talking to you.
- `msr-flows/Notes/.ipynb_checkpoints` is cruft and should be removed from this PR
- There is some inconsistency with how you write docstrings in your `.py` files. For example, in the `_init_` function for the `full_run_serp` class in `run_user_input.py`, you have the following sections
   ```
   Returns
   -------
   None
   ```
   But later in `linear_generation_reprocessing_constants`, you have
   ```
   Returns
   -------
   
   ```
   and then in `cycle_time_decay`, there isn't a `Returns` section in the docstring at all. Ideally, all of your function docstrings would have a `Returns` section with a full description of the returned values, or `None` if no values are returned. Similar inconsistencies exist in the use of the `Parameters` section, and in the use of docstrings at all. Consider going back through your docstrings in all of your `.py` files and make the necessary edits to ensure they're all consistent.


## Reviewer checkist
- [x] Read the PR description
- [x] Read through all the changes, considering the following questions.
  - [x] Is there anything you can praise about this PR? Start with
        praise.
  - [x] Are variable names brief but descriptive?
  - [ ] Are new/changed functions no longer than a paragraph?
  - [ ] Do all function parameters have default values where appropriate?
  - [ ] Is the code clear and clean? (see Robert C. Martin's
        [Clean Code](https://i-share-uiu.primo.exlibrisgroup.com/discovery/fulldisplay?docid=alma99944155312205899&context=L&vid=01CARLI_UIU:CARLI_UIU&tab=LibraryCatalog&lang=en))
  - [x] Is there enough documentation?
  - [ ] Does the programming style meet the requirements of the
        repository ([PEP8](https://www.python.org/dev/peps/pep-0008/) for python,
        [google C++ style guide](https://google.github.io/styleguide/cppguide.html), etc.)
  - [ ] If there is no continuous integration, check out the branch
        locally, build, and run the tests.
  - [ ] Do the tests pass on your machine?
  - [ ] Is the PR free of random cruft (built files, `.swp` files,
        etc.)?
  - [ ] Do all files in the PR belong in the repository?
  - [ ] If the PR deletes files, is this appropriate?
  - [ ] If the PR adds files or data, are these new files compatible
        with the repository license?
  - [x] Make a review, leaving kind comments and suggesting changes
        where needed (to resolve the above).
  - [ ] Has the author resolved all of the comments and suggestions
        in your review?
- [ ] When you approve of the PR, merge and close it.
- [ ] Does this PR close an issue? If so, be sure to descriptively
      close this issue when the PR is merged
- [x] Thank the author for their contribution.

**Comment by LukeSeifert** [comment]
@yardasol Thanks for the reviewing the PR! I'm pretty busy this week, but I should be able to get around to your comments next week or the week after.

File: msr-flows/readme.md
Comment (LukeSeifert) [reply msr-flows/readme.md:5]: Yeah, I was using the original scripts for testing different things, so the refactor also trimmed some of those.

File: msr-flows/msbr_compare/serpent_output.py
Comment (LukeSeifert) [reply msr-flows/msbr_compare/serpent_output.py:1]: I wasn't aware of that actually, I'm just used to using Serpent Tools. I might have to look into the `pyne` version.

**Comment by LukeSeifert** [comment]
I originally had this work on its own branch before I merged it with master, that makes sense to PR the branch I was working on. I don't think I'm able to change it without closing this PR and opening a new one, but I will definitely keep it in mind for next time. 

I believe I addressed most of the suggestions you made, but the pep8 issues weren't fixable with autopep8 for some reason (I think because a lot of them were doc strings), so I did those manually. For the `saltproc_read.py`, since its a script I copied from SaltProc just for ease of access, I wasn't sure if I should make changes to the built-in functions, that's why I just added my own function to it and didn't change anything else. In the future, I should be able to structure things slightly differently such that I won't need to use that script anymore though, so I think it is only a temporary issue.

File: msr-flows/readme.md
Comment (yardasol) [reply msr-flows/readme.md:5]: Okay. Marking as resolved

**Comment by yardasol** [comment]
Looks like all your pep8 issues have been resolved. Nice work.
You should really add docstrings for the parameters in `get_library_isotopes` and `check_isotope_in_library`. Other than that, everything else looks good.

File: msr-flows/msbr_compare/serpent_input.py
Comment (yardasol) [inline msr-flows/msbr_compare/serpent_input.py:33]: You should add what the keys and values for the dictionary like this:

```
        reprocessing : dict
            Dictionary that...

            ``key``
                key description
            ``value``
                value description

```
do this for all `dict` parameters in your function docstrings.

**Review by yardasol — CHANGES_REQUESTED** [review]
Hi @LukeSeifert, I've looked at your changes and I like how thorough you were with the additions to the docstrings.

I noticed a few typos that have suggestions ready for commit.

I brought up a few points in the in-line comments that I think are worth repeating here, in brief at least:
- Every docstring is now bounded by a `'''` instead of `"""`. `"""` is the correct format. Is there a reason you changed this?
- Make sure to add `key` and `value` descriptions to parameters of type `dict` in your docstrings. I've given you an example of how to do this in the in-line comment.

Sorry for being such a stickler about this. I think it'll save time in the long run to have these smaller pieces taken care of sooner rather than later (and doing these fixes now preps you for making those additions to your coding workflow in the future!)

**Comment by LukeSeifert** [comment]
Thanks for the comments! I did not realize that the triple quote is used over the triple apostrophe, so I went through and replaced that in the other files. I also added the proper dictionary docstrings and fixed the typos you pointed out.

File: msr-flows/msbr_compare/serpent_calculations.py
Comment (LukeSeifert) [reply msr-flows/msbr_compare/serpent_calculations.py:100]: I think this should be alright as it is since there is some more context in the rest of the docstring now that I fixed the dictionary docstrings.

**Review by yardasol — APPROVED** [review]
Good work. I didn't see anything else looking over the code, so I think you can go ahead an merge!

**Pull request merged**
